source/slang/slang-ir-transform-params-to-constref.cpp

```diff
@@ -0,0 +1,321 @@
+            return true;
+
+        // Skip functions with other shader stage decorations
+        if (func->findDecoration<IRMaxVertexCountDecoration>() ||
+            func->findDecoration<IRInstanceDecoration>() ||
+            func->findDecoration<IRWaveSizeDecoration>() ||
+            func->findDecoration<IRGeometryInputPrimitiveTypeDecoration>() ||
+            func->findDecoration<IRStreamOutputTypeDecoration>())
+            return false;
+
+        // Skip functions with `kIROp_GenericAsm` since
+        // these instructions inject target specific code
+        // using parameters in an unpredictable way, relying
+        // on assumptions that parameters do not change type.
+        for (auto block : func->getBlocks())
+        {
+            for (auto inst : block->getChildren())
+            {
+                if (!as<IRGenericAsm>(inst))
+                    continue;
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    // Process a single function
+    void processFunc(IRFunc* func)
+    {
+        HashSet<IRParam*> updatedParams;
+        bool hasTransformedParams = false;
+
+        // First pass: Transform parameter types
+        for (auto param = func->getFirstParam(); param; param = param->getNextParam())
+        {
+            if (shouldTransformParam(param))
+            {
+                // Our goal here is to transform `in T` parameters to const-ref.
+                // We are selective about what we will transform for a few reasons:
+                // 1. no reason to transform simple primitives like `int`.
+                // 2. not every type makes sense as constref. For example, `ParameterBlock`.
+                // 3. constref is not 100% stable, so we need to be selective on what we let
+                //    transform into constref.
+                //
+                // This allows us to pass the address of variables directly into a function,
+                // giving us the choice to remove copies into a parameter.
+                auto paramType = param->getDataType();
+                auto constRefType = builder.getConstRefType(paramType);
+                param->setFullType(constRefType);
+
+                hasTransformedParams = true;
+                changed = true;
+                updatedParams.add(param);
+            }
+        }
+
+        if (!hasTransformedParams)
+        {
+            return;
+        }
+
+        // Second pass: Update function body according to the new `constref` parameters
+        rewriteParamUseSitesToSupportConstRefUsage(updatedParams);
+
+        // Third pass: Update call sites
+        updateCallSites(func, updatedParams);
+    }
+
+    SlangResult processModule()
+    {
+        // Collect all functions that need processing
+        List<IRFunc*> functionsToProcess;
+
+        for (auto inst = module->getModuleInst()->getFirstChild(); inst; inst = inst->getNextInst())
```

Comment:
I think we should process the callgraph in topological order, so callees are always processed before callers.